Blog post: Stable Ember TypeScript types (#1285)
ember-learn/ember-blog: content/stable-typescript-types-in-ember-5-1.md

- [Getting started](#toc_getting-started)
    - [JavaScript](#toc_javascript)
    - [TypeScript](#toc_typescript)
- [Caveats](#toc_caveats)
- [Stability](#toc_stability)
- [Migration Guide](#toc_migration-guide)
    - [From the preview types](#toc_from-the-preview-types)
    - [From DefinitelyTyped](#toc_from-definitelytyped)
- [Background](#toc_background)
- [Onward](#toc_onward)

## Getting started

Both JavaScript and TypeScript developers can benefit from using these new type definitions. If you are JavaScript developer, this will make automatic imports, in-editor framework documentation, and other features work much better for you in your Ember app or addon. If you are a TypeScript user, you will get all of those benefits as well as the strong type safety guarantees and refactoring capabilities you are used to.

### JavaScript

In your `app/app.js`, add this single doc comment to the bottom of the file:

```js
/**
 * @typedef {import('ember-source/types')} EmberTypes
 */
```

> Now you will get autocomplete and in-editor documentation for all the `@ember/*` packages which ship with Ember with no other adjustment required!

Reviewer: Is this true regardless of IDE or only for IDEs like VSCode that come with baked-in TS support?

Author: Ah! Good catch. We should clarify that it will be true *for any editor configured to use the TS LS*. I will update to note that on Monday!

Author: ```suggestion
This is all that is necessary to unlock autocomplete and in-editor documentation for all the `@ember/*` packages which ship with Ember with no other adjustment required! Any editor configured to use the TypeScript language server for JavaScript support will "just work" once you make this change—including VS Code, where this works in a default installation.
```

For details on migrating existing Ember TypeScript projects, see the [Migration Guide](#toc_migration-guide) below!

That's all you need to do to enjoy the new TypeScript support!

## Caveats

However, there are a couple important caveats you should know about.

1. These new types are not compatible with the existing `@types` definitions for Ember Data. Unfortunately, Ember Data's hand-written type definitions rely in a number of key areas on some of the mistakes in the `@types` packages for Ember—mistakes that are fixed by using types published from Ember's own source code. So far, the Ember Data team has had too few contributors to make fixing this a top priority. We made the choice to go ahead and publish these stable types from Ember anyway for two reasons:

    - While many Ember users *do* rely on Ember Data, not all do, and this makes the best possible types for Ember available to anyone who is not blocked in this way.

    - Ember Data itself can now take advantage of the Ember types! This was a long-standing chicken-or-egg problem: Ember Data could not really publish its own stable types while relying on the hand-written `@types` packages for Ember!

>     Happily, the two newest members of the Ember TypeScript team, Peter Wagenet and Krystan HuffMenne, are both actively interested in tackling this problem, so there will hopefully be good progress in this space in the rest of 2023. (This is a great place to chip in and help, by the way!)

Reviewer: 🙈

## Stability

Because of the interop challenge with Ember Data, our TypeScript support is *Released* but not yet *Recommended*. This is one of the first major features to land in Ember which leans on this distinction in a way that matters to end users, so here is a quick explainer. Since [RFC 0617: RFC Stages][rfc-0617], features go through up to six stages on the way to being “done”:

0. **Proposed:** A proposal for a change to Ember or its processes that is offered for community and team evaluation.
1. **Exploring:** An RFC deemed worth pursuing but in need of refinement.
2. **Accepted:** A fully specified RFC. Waiting for or in the process of implementation.
3. **Ready for Release:** The implementation of the RFC is complete, including learning materials.
4. **Released:** The work is published. If it is codebase-related work, it is in a stable version of the relevant package(s).
5. **Recommended:** The feature/resource is recommended for general use.

> Since TypeScript is *Released*, you can rely on all of our normal SemVer guarantees. Ember's types are now part of our public contract with you as a user, and they have to be maintained with the exact same care as any other part of the framework. Additionally, Chris Krycho and the rest of the Ember TypeScript team developed [a spec for Semantic Versioning for TypeScript Types][semver-ts], which allows Ember (and any other framework or library!) to define how it manages supported TypeScript versions and absorb breaking changes from TypeScript. That means it is totally safe for you to rely on these types if you are not impacted by the caveats described above!

Author: ```suggestion
Since TypeScript is *Released*, you can rely on all of our normal SemVer guarantees. Ember's types are now part of our public contract with you as a user, and they have to be maintained with the exact same care as any other part of the framework. Additionally, Chris Krycho and the rest of the Ember TypeScript team developed [a spec for Semantic Versioning for TypeScript Types][semver-ts], which allows Ember (and any other framework or library) to define how it manages supported TypeScript versions and absorb breaking changes from TypeScript. That means it is totally safe for you to rely on these types if you are not impacted by the caveats described above!
```

> 

Author: ```suggestion

One specific point: as described on our [Releases](https://emberjs.com/releases/) page, the Ember framework itself uses a “rolling window” support policy for TypeScript versions. This means Ember can bump its minimum supported TypeScript version in an Ember minor release, just like it can for Node.js versions—but the supported range of TypeScript versions across two Ember LTS releases must always overlap. So if Ember 5.4 LTS supports TypeScript 4.9–5.3, Ember 5.8 LTS *could* drop support for TypeScript versions earlier than TypeScript 5.3, but must include at least TypeScript 5.3. That guarantees that users can always upgrade their TypeScript and Ember versions separately. In practice, we expect to maintain a fairly wide TypeScript version support window to minimize the impact on the ecosystem—likely (though not guaranteed to be) 4 to 6 recent TypeScript versions at any given time.

Ember's initial supported TypeScript version range is TypeScript 4.9, 5.0, and 5.1.

```

> The only things that will change once Ember Data publishes types which are compatible with these and we can thus move to “Recommended” are:

Author: ```suggestion
The only other things that will change once Ember Data publishes types which are compatible with these and we can thus move to “Recommended” are:
```

    - [From the preview types](#toc_from-the-preview-types)
    - [From DefinitelyTyped](#toc_from-definitelytyped)
- [Background](#toc_background)
- [Onward](#toc_onward)

## Getting started

Both JavaScript and TypeScript developers can benefit from using these new type definitions. If you are JavaScript developer, this will make automatic imports, in-editor framework documentation, and other features work much better for you in your Ember app or addon. If you are a TypeScript user, you will get all of those benefits as well as the strong type safety guarantees and refactoring capabilities you are used to.

### JavaScript

In your `app/app.js`, add this single doc comment to the bottom of the file:

```js
/**
 * @typedef {import('ember-source/types')} EmberTypes
 */
```

This is all that is necessary to unlock autocomplete and in-editor documentation for all the `@ember/*` packages which ship with Ember with no other adjustment required! Any editor configured to use the TypeScript language server for JavaScript support will "just work" once you make this change—including VS Code, where this works in a default installation.

### TypeScript

> For a new TypeScript project, the setup is similar, but has some additional steps for now. In the future, these steps will not be necessary. For now, the default TypeScript blueprint continues to set up users for maximum backward compatibility with the rest of the Ember ecosystem.

Author: ```suggestion
Using the stable types in a new TypeScript projects is similar to setting it up for JavaScript, but has a few additional steps. To provide maximum backward compatibility with the rest of the Ember ecosystem, the default TypeScript blueprint still uses the `@types` packages, and configures those packages to use Array prototype extensions. Both of these are incompatible with the stable types, so we need to remove them! In the future, once we can switch to defaulting to using the stable types in the default blueprint, these extra steps will not be necessary.

(For details on migrating *existing* Ember TypeScript projects, see the [Migration Guide](#toc_migration-guide) below as well.)

```

  interface Array<T> extends Ember.ArrayPrototypeExtensions<T> {}
  // interface Function extends Ember.FunctionPrototypeExtensions {}
}

export {};
```

You can replace that whole gnarly mess with just this:

```ts
import 'ember-source/types';
```

Second, you may need to update to recent versions of `ember-qunit`, `ember-resolver`, and `@ember/test-helpers`, which have themselves been updated to use Ember's native types.

Finally, you will need to remove the following packages from your `package.json` file and reinstall your package manager's installation command to update your `node_modules`:

- `@types/ember`
- `@types/ember__application`
- `@types/ember__array`
- `@types/ember__component`
- `@types/ember__controller`
- `@types/ember__debug`
- `@types/ember__destroyable`
- `@types/ember__engine`
- `@types/ember__error`
- `@types/ember__helper`
- `@types/ember__modifier`
- `@types/ember__object`
- `@types/ember__owner`
- `@types/ember__polyfills`
- `@types/ember__routing`
- `@types/ember__runloop`
- `@types/ember__service`
- `@types/ember__string`
- `@types/ember__template`
- `@types/ember__test`
- `@types/ember__utils`

> For details on migrating existing Ember TypeScript projects, see the [Migration Guide](#toc_migration-guide) below!
> 

Author: ```suggestion
```
